Expand use of WKBundlePagePostMessageWithAsyncReply in WebKitTestRunner

achristensen07 opened the pull request:
#### 453783344f766c0d7498541a692c0e9a1951f5fc
<pre>
Expand use of WKBundlePagePostMessageWithAsyncReply in WebKitTestRunner
<a href="https://bugs.webkit.org/show_bug.cgi?id=272593">https://bugs.webkit.org/show_bug.cgi?id=272593</a>
<a href="https://rdar.apple.com/126361971">rdar://126361971</a>

Reviewed by Sihui Liu and Charlie Wolfe.

This makes tests work better with site isolation because the reply is sent to the
process that made the request, instead of always sending the reply to the main frame&apos;s process.

Most of this PR is just plumbing to get the call in the UI process, get the reply, send it
to the web process, and process it into a form that JS can receive in the web process.  I also
reduced duplicate implementations of TestController::getAllStorageAccessEntries.

I verified this works correctly by running the 1 test that uses getAllStorageAccessEntries:
http/tests/storageAccess/request-and-grant-access-then-detach-should-not-have-access.html

* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::didReceiveMessageToPage):
(WTR::asyncReplyHandler):
(WTR::postMessageWithAsyncReply):
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.h:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::stringArrayToJS):
(WTR::TestRunner::getAllStorageAccessEntries):
(WTR::TestRunner::callDidReceiveAllStorageAccessEntriesCallback): Deleted.
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::createOtherPlatformWebView):
(WTR::TestController::createWebViewWithOptions):
(WTR::TestController::didReceiveAsyncMessageFromInjectedBundle):
(WTR::getAllStorageAccessEntriesCallback):
(WTR::TestController::getAllStorageAccessEntries):
(WTR::GetAllStorageAccessEntriesCallbackContext::GetAllStorageAccessEntriesCallbackContext): Deleted.
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):
(WTR::TestInvocation::didReceiveAllStorageAccessEntries): Deleted.
* Tools/WebKitTestRunner/TestInvocation.h:
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::getAllStorageAccessEntries): Deleted.

Canonical link: <a href="https://commits.webkit.org/277505@main">https://commits.webkit.org/277505@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4e852f89d055a62f98342ea5eab818abf2affb4b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47813 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27006 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50624 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50495 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43867 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50119 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32876 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24492 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38914 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48395 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/24681 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41320 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20206 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/22159 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/42513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5861 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/44172 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/42941 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52389 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22850 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19202 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/46222 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24121 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45259 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/24911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6762 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23843 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
```diff
@@ -60,6 +60,7 @@
 #import <WebKit/_WKWebsiteDataStoreConfiguration.h>
 #import <pal/spi/cocoa/LaunchServicesSPI.h>
 #import <wtf/BlockPtr.h>
+#import <wtf/CompletionHandler.h>
```
Doesn't seem this like is needed?

charliewolfe commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -1819,6 +1819,11 @@ void TestController::didReceiveSynchronousPageMessageFromInjectedBundleWithListe
     static_cast<TestController*>(const_cast<void*>(clientInfo))->didReceiveSynchronousMessageFromInjectedBundle(messageName, messageBody, listener);
 }
 
+void TestController::didReceiveAsyncMessageFromInjectedBundle(WKPageRef page, WKStringRef messageName, WKTypeRef messageBody, WKMessageListenerRef listener, const void* clientInfo)
```
nit - including "PageMessageFromInjectedBundleWithListener" in the function name would match the function name of the sync version above. also `page` is an unused parameter.

szewai commented on Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
```diff
@@ -1086,27 +1068,38 @@ static JSContextRef firstRootFrameJSContext()
+        JSValueRef* arguments { nullptr };
+        JSValueRef resultJS { nullptr };
+
+        if (result) {
```
result can be null when message is `GetAllStorageAccessEntries` and `!page`. In this case it seems we should set argumentCount to 1 and argument to an empty or null `JSValue`.
We should probably decide the arguments by message name.

achristensen07 replied:
I think I disagree.  Calling JSObjectCallAsFunction with 1 argument that is null or 0 arguments is pretty equivalent, and doing so would make the abstraction postMessageWithAsyncReply less useful.  As we continue to expand the use of postMessageWithAsyncReply, though, we may find that you're right and we can do something about it then.

szewai replied:
Ok if they are equivalent then it's fine. Yeah my idea is generally performing the type check based on what message it is, to make the code more robust.

achristensen07 replied:
That would make it more robust, but also more verbose.  I think it's fine to have an increasingly generic sendWithAsyncReply wrapper that takes whatever WKTypeRef is in the reply and turn it into a JS object, but I'll keep this in mind, and I'll consider whether the more verbose code would be worth it.  This is just a simple case, so it's hard to tell before more widespread adoption.

achristensen07 committed:
Expand use of WKBundlePagePostMessageWithAsyncReply in WebKitTestRunner
https://bugs.webkit.org/show_bug.cgi?id=272593
rdar://126361971

Reviewed by Sihui Liu and Charlie Wolfe.

This makes tests work better with site isolation because the reply is sent to the
process that made the request, instead of always sending the reply to the main frame's process.

Most of this PR is just plumbing to get the call in the UI process, get the reply, send it
to the web process, and process it into a form that JS can receive in the web process.  I also
reduced duplicate implementations of TestController::getAllStorageAccessEntries.

I verified this works correctly by running the 1 test that uses getAllStorageAccessEntries:
http/tests/storageAccess/request-and-grant-access-then-detach-should-not-have-access.html

* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::didReceiveMessageToPage):
(WTR::asyncReplyHandler):
(WTR::postMessageWithAsyncReply):
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.h:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::stringArrayToJS):
(WTR::TestRunner::getAllStorageAccessEntries):
(WTR::TestRunner::callDidReceiveAllStorageAccessEntriesCallback): Deleted.
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::createOtherPlatformWebView):
(WTR::TestController::createWebViewWithOptions):
(WTR::TestController::didReceiveAsyncMessageFromInjectedBundle):
(WTR::getAllStorageAccessEntriesCallback):
(WTR::TestController::getAllStorageAccessEntries):
(WTR::GetAllStorageAccessEntriesCallbackContext::GetAllStorageAccessEntriesCallbackContext): Deleted.
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):
(WTR::TestInvocation::didReceiveAllStorageAccessEntries): Deleted.
* Tools/WebKitTestRunner/TestInvocation.h:
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::getAllStorageAccessEntries): Deleted.

Canonical link: https://commits.webkit.org/277505@main
